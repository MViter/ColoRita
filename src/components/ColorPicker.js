import React from 'react'
import { RgbaColorPicker } from "react-colorful";
import { HeaderTitle, PaletteContainer } from '../styles.js';

const ColorPicker = ({ color, setColor }) => {
    return (<PaletteContainer>
        <HeaderTitle className="text">React Colorful ðŸŽ¨</HeaderTitle>
        <RgbaColorPicker data-testid="color-picker" color={color} onChange={setColor} />
    </PaletteContainer>)
}

export default ColorPicker