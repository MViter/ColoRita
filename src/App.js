import React, { useState, useEffect, useRef, useCallback } from 'react'
import { TextDescription, StyledButton, CenteredContainer } from './styles.js';
import { useBodyBackground } from "./hooks"
import ColorItem from './components/ColorItem'
import ColorPicker from './components/ColorPicker'
import MultipleColorsShema from './components/MultipleColorsShema'
import ColorForm from './components/ColorForm'
import { getRGBAColorString, getTextColor, RGBAToHexA} from './utils'

// https://github.com/omgovich/react-colorful
// https://www.thecolorapi.com/docs#colors-color-identification-get
// https://goober.js.org/

function App() {
  const initialColor = { r: 255, g: 255, b: 255, a: 1 }
  const [colors, setColors] = useState([initialColor])
  const [showSchema, setShowScema] = useState(false)
  const [showCustomInput, setShowCustomInput] = useState(false)
  const count = useRef()

  const setColor = useCallback(color => {
    clearTimeout(count.current)
    count.current = setTimeout(() => {
      setColors([...colors, color])
    }, 500)
  }, [colors])

  useEffect(() => {
    return () => clearTimeout(count.current)
  }, [])

  useBodyBackground(getRGBAColorString(colors[colors.length - 1]), getTextColor(colors[colors.length - 1]))

  // console.log('Re-render App');

  const switchShowScema = () => setShowScema(!showSchema)

  const switchShowCustomInput = () => setShowCustomInput(!showCustomInput)

  return (
    <div className="main-app">
      <div className="app-content">
        <CenteredContainer>
          <ColorPicker color={colors[colors.length - 1]} setColor={setColor} />
          <StyledButton className="button-text" onClick={switchShowScema} bgColor={RGBAToHexA(getRGBAColorString(colors[colors.length - 1]))} data-testid="show-schema">Show color schema</StyledButton>
          <StyledButton className="button-text" onClick={switchShowCustomInput} bgColor={RGBAToHexA(getRGBAColorString(colors[colors.length - 1]))} data-testid="add-color">Add custom color</StyledButton>
          {showSchema && <MultipleColorsShema color={colors[colors.length - 1]} />}
          {showCustomInput && <ColorForm setColor={setColor} color={colors[colors.length - 1]} />}
          <TextDescription className="text">
            Your colors:
          </TextDescription>
        </CenteredContainer>
        <div className="colors">
          {
            colors.map((c, index) => <ColorItem index={index} color={c} key={index} />)
          }
        </div>
        
      </div>
    </div>
  )
}


export default App
