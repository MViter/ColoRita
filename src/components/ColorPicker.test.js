import React, { useState } from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
// import user from '@testing-library/user-event'
//import fireEvent from '@testing-library/user-event'
import ColorPicker  from './ColorPicker';
import App from '../App';

const mock = jest.fn()
const mockSetColor = jest.fn()
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn(() => [{}, () => []])
}))


beforeEach(() => {
    mock.mockClear()
    mockSetColor.mockClear()
    useState.mockImplementation(() => [{}, mockSetColor])
})

describe('ColorPicker suite', () => {
  test('snapshop test of the ColorPicker', () => {
    // useState.mockImplementation(() => [changedColor, mockSetColor])
    const { getByText, queryByText, baseElement } = render(<ColorPicker color={'#00FF00'} setColor={mock}/>)
    expect(queryByText('erterer')).not.toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  test('verify ColorPicker is in document', () => {
    const { getByTestId } = render(<ColorPicker color={'#00FF00'} setColor={mock}/>)
    expect(getByTestId('color-picker')).toBeInTheDocument()
  })

  test('verify setColor will be called on ColorPicker change', () => {
    // useState.mockImplementation(() => [changedColor, mockSetColor])
    
    // const { baseElement, getByTestId } = render(<ColoPicker color={'#00FF00'} setColor={mock}/>)
    const { getByTestId } = render(<App color={'#00FF00'} setColor={mock}/>)
    fireEvent.change(getByTestId('color-picker'), { target: {r: 255, g: 23, b: 23, a: 1}});
    expect(mockSetColor).toHaveBeenCalledWith({"a": "1", "b": "", "g": "", "r": correctValue2})
  })
})