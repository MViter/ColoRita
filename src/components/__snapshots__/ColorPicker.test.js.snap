// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker suite snapshop test of the ColorPicker 1`] = `
<body>
  <div>
    <span
      class="sc-hLseeU cgmizj"
    >
      <h3
        class="sc-beqWaB kcKTaF text"
      >
        React Colorful ðŸŽ¨
      </h3>
      <div
        class="react-colorful"
        data-testid="color-picker"
      >
        <div
          class="react-colorful__saturation"
        >
          <div
            aria-label="Color"
            aria-valuetext="Saturation NaN%, Brightness NaN%"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__saturation-pointer"
            >
              <div
                class="react-colorful__pointer-fill"
              />
            </div>
          </div>
        </div>
        <div
          class="react-colorful__hue"
        >
          <div
            aria-label="Hue"
            aria-valuemax="360"
            aria-valuemin="0"
            aria-valuenow="NaN"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__hue-pointer"
              style="top: 50%;"
            >
              <div
                class="react-colorful__pointer-fill"
              />
            </div>
          </div>
        </div>
        <div
          class="react-colorful__alpha react-colorful__last-control"
        >
          <div
            class="react-colorful__alpha-gradient"
          />
          <div
            aria-label="Alpha"
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="NaN"
            aria-valuetext="NaN%"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__alpha-pointer"
              style="top: 50%;"
            >
              <div
                class="react-colorful__pointer-fill"
              />
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
</body>
`;
